### PR TITLE
feat: session space

### PR DESCRIPTION
**Pull request by huaxiabuluo**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

### CODE REVIEW

File: server/api/studio/pkg/ws/client.go [hunk 1]
```diff
@@ -55,17 +57,39 @@ var upgrader = websocket.Upgrader{
...
 
+func (c *Client) switchSpace(msgReceived *MessageReceive) *map[string]any {
+	reqSpace, ok := msgReceived.Body.Content["space"].(string)
+	shouldSwitch := ok && reqSpace != "" && nsidSpaceMap[c.clientInfo.NSID] != reqSpace
+	if !shouldSwitch {
+		return nil
+	}
+	// name.replace(/\\/gm, '\\\\').replace(/`/gm, '\\`')
+	reqSpace = strings.Replace(reqSpace, "\\", "\\\\", -1)
+	reqSpace = strings.Replace(reqSpace, "`", "\\`", -1)
```
Comment (hetao92) [inline server/api/studio/pkg/ws/client.go:76]: need to add  `reqSpace` ?

File: server/api/studio/pkg/ws/client.go [hunk 2]
```diff
@@ -55,17 +57,39 @@ var upgrader = websocket.Upgrader{
...
+		content := map[string]any{
+			"code":    base.Error,
+			"message": err.Error(),
+		}
+		if auth.IsSessionError(err) {
+			content["code"] = ecode.ErrSession.GetCode()
+		}
+		return &content
+	}
+	nsidSpaceMap[c.clientInfo.NSID] = reqSpace
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: Concurrency conflicts.

File: server/api/studio/pkg/ws/client.go [hunk 3]
```diff
@@ -231,6 +274,7 @@ func (c *Client) readPump() {
 func (c *Client) writePump() {
 	ticker := time.NewTicker(pingPeriod)
 	defer func() {
+		delete(nsidSpaceMap, c.clientInfo.NSID)
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go]: Concurrency conflicts.

File: server/api/studio/pkg/ws/client.go (hunk 2 above)
Comment (huaxiabuluo) [reply server/api/studio/pkg/ws/client.go]: done

File: server/api/studio/pkg/ws/client.go (hunk 3 above)
Comment (huaxiabuluo) [reply server/api/studio/pkg/ws/client.go]: done

File: server/api/studio/pkg/utils/tools.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package utils
+
+import "sync"
+
+type MutexMap[T any] struct {
+	mu   sync.RWMutex
+	data map[string]T
```
Comment (veezhang) [inline server/api/studio/pkg/utils/tools.go:7]: Panic without init.

File: server/api/studio/pkg/ws/client.go [hunk 5]
```diff
@@ -37,35 +39,49 @@ const (
...
 }
 
+func (c *Client) switchSpace(msgReceived *MessageReceive) *map[string]any {
+	reqSpace, ok := msgReceived.Body.Content["space"].(string)
+	currentSpace, _ := nsidSpaceMap.Get(c.clientInfo.NSID)
+
+	shouldSwitch := ok && reqSpace != "" && currentSpace != reqSpace
+	if !shouldSwitch {
+		return nil
+	}
```
Comment (veezhang) [inline server/api/studio/pkg/ws/client.go:64]: Why not `if !ok || reqSpace == "" ||  currentSpace == reqSpace { return nil}`?

Comment (huaxiabuluo) [reply server/api/studio/pkg/ws/client.go:64]: make it more semantic

**Review by hetao92 — APPROVED** [review]
LGTM

**Pull request merged**